src/Data/Singletons/Single/Defun.hs

```diff
@@ -0,0 +1,220 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.Singletons.Single.Defun
+-- Copyright   :  (C) 2018 Ryan Scott
+-- License     :  BSD-style (see LICENSE)
+-- Maintainer  :  Richard Eisenberg (rae@cs.brynmawr.edu)
+-- Stability   :  experimental
+-- Portability :  non-portable
+--
+-- Creates 'SingI' instances for promoted types' defunctionalization symbols.
+--
+-----------------------------------------------------------------------------
+
+module Data.Singletons.Single.Defun (singDefuns) where
+
+import Data.List
+import Data.Singletons.Names
+import Data.Singletons.Promote.Defun
+import Data.Singletons.Single.Monad
+import Data.Singletons.Single.Type
+import Data.Singletons.Util
+import Language.Haskell.TH.Desugar
+import Language.Haskell.TH.Syntax
+
+-- Given the Name of something, take the defunctionalization symbols for its
+-- promoted counterpart and create SingI instances for them. As a concrete
+-- example, if you have:
+--
+--   foo :: Eq a => a -> a -> Bool
+--
+-- Then foo's promoted counterpart, Foo, will have two defunctionalization
+-- symbols:
+--
+--   FooSym0 :: a ~> a ~> Bool
+--   FooSym1 :: a -> a ~> Bool
+--
+-- We can declare SingI instances for these two symbols like so:
+--
+--   instance SEq a => Sing (FooSym0 :: a ~> a ~> Bool) where
+--     sing = singFun2 sFoo
+--
+--   instance (SEq a, SingI x) => SingI (FooSym1 x :: a ~> Bool) where
+--     sing = singFun1 (sFoo (sing @_ @x))
```

Comment:
This business of always having to specify the kind as `_` kind of stinks, especially since we don't have to do this for the closely related function [`demote`](https://hackage.haskell.org/package/singletons-2.4.1/docs/Data-Singletons.html#v:demote). Perhaps we should change `SingI` to:

```haskell
class SingI a where
  sing :: a
```

So that the kind of `a` becomes invisible?